sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -63,10 +63,27 @@ object PhysicalOperation extends PredicateHelper {
         val substitutedFields = fields.map(substitute(aliases)).asInstanceOf[Seq[NamedExpression]]
         (Some(substitutedFields), filters, other, collectAliases(substitutedFields))
 
-      case Filter(condition, child) if condition.deterministic =>
-        val (fields, filters, other, aliases) = collectProjectsAndFilters(child)
-        val substitutedCondition = substitute(aliases)(condition)
-        (fields, filters ++ splitConjunctivePredicates(substitutedCondition), other, aliases)
+      case filter: Filter if filter.condition.deterministic || filter.condition.isInstanceOf[And] =>
+        val condition = filter.condition
+        val determinedCondition = if (condition.deterministic) {
+          Some(condition)
+        } else {
+          val andCondition = condition.asInstanceOf[And]
+          if (andCondition.left.deterministic) {
```

Comment:
Do we intend to handle only at the top level ? What happens when the non-deterministic predicate is nested like : 
```
where ((c1 = 10 and rand() < 1) and (c2 = 20))
```
in this case, in my understanding of the code, we will ignore (c1 = 10) for pruning purposes ? cc @maropu

Reply:
yea, we need more general solution for that..

Reply:
@dilipbiswal @maropu It has been updated. Please review it, thanks~~